### PR TITLE
[BL] Adding vertical integration kernels

### PR DESCRIPTION
**Pull request by jkozdon**
Adds functionality to compute indefinite line integrals for stacked meshes along the stacked dimension though the function `DGBalanceLawDiscretizations.indefinite_stack_integral!`.

The function `preodefun!` is also introduced which allows the user to "inject" code right before the ODE function is evaluated.

Few other bug/feature fixes:
- Read geo indices from Grids and unify names
- Removing unused functions in MPIStateArrays
- Remove some structure assumptions in MPIStateArrays
- Fix bug Topology for stacksize and getproperty
- Fix `grad_test_sphere.jl` to actually use a sphere

### CODE REVIEW

**Comment by jkozdon** [comment]
cc: @lcw @smarras79 @asraero

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

File: src/DGmethods/DGBalanceLawDiscretizations.jl [hunk 1]
```diff
@@ -659,6 +688,64 @@ function grad_auxiliary_state!(disc::DGBalanceLaw, id, (idx, idy, idz))
                            Dmat, topology.elems, id, idx, idy, idz))
 end
 
+
+"""
+    indefinite_stack_integral!(disc, f, Q, out_states, [P=disc.auxstate])
+
+Computes the indefinite integral along an element stack using state `Q`
+```math
+∫_{ζ_{0}}^{ζ} f(q; aux, t)
```
Comment (simonbyrne) [inline src/DGmethods/DGBalanceLawDiscretizations.jl]: Is this the integral from the base upwards? I think in some cases they need the integral in the other direction

File: src/DGmethods/DGBalanceLawDiscretizations.jl [hunk 2]
```diff
@@ -659,6 +688,64 @@ function grad_auxiliary_state!(disc::DGBalanceLaw, id, (idx, idy, idz))
                            Dmat, topology.elems, id, idx, idy, idz))
 end
 
+
+"""
+    indefinite_stack_integral!(disc, f, Q, out_states, [P=disc.auxstate])
```
Comment (simonbyrne) [inline src/DGmethods/DGBalanceLawDiscretizations.jl]: should be added to the relevant `docs/src/` file.

File: src/DGmethods/DGBalanceLawDiscretizations.jl [hunk 3]
```diff
@@ -659,6 +688,64 @@ function grad_auxiliary_state!(disc::DGBalanceLaw, id, (idx, idy, idz))
...
+where `F` is an `MVector` of length `length(out_states)`, `Q` and `aux` are
+the `MVectors` for the state and auxiliary state at a single degree of freedom.
+The function is responsible for filling `F`.
+
+Requires the `isstacked(disc.grid.topology) == true`
+
+!!! note
+
+    When the topology has one two dimensions the integral is computed along the
+    `η` direction.
```
Comment (simonbyrne) [inline src/DGmethods/DGBalanceLawDiscretizations.jl]: would it be simplest to say "trailing dimension"?

File: src/DGmethods/DGBalanceLawDiscretizations.jl (hunk 1 above)
Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: Yes it is. Given the top value you can extract the other one... that said we do not have convenient way to access this at the user level.

Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: Could change this to:
```julia
indefinite_stack_integral!(disc::DGBalanceLaw, f, Q, out_states,
                                    P=disc.auxstate; direction=:ascending)
```
or something like that.

File: src/DGmethods/DGBalanceLawDiscretizations.jl (hunk 3 above)
Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: sure.

File: src/DGmethods/DGBalanceLawDiscretizations.jl (hunk 2 above)
Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: done (as well as fixed other docs issues)

File: src/DGmethods/DGBalanceLawDiscretizations.jl (hunk 1 above)
Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: How would one `accumulate` backward in julia? This would really be the analog in summation case.

Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: One option would be to have a reverse a precomputed indefinite integral function if up is the standard way, but sometimes backward is also needed.

**Comment by jkozdon** [comment]
bump.

@asraero and @smarras79 do you guys need anything else in this functionality?

Also, rebased this on master (since there were going to be conflicts)

File: src/DGmethods/DGBalanceLawDiscretizations.jl (hunk 1 above)
Comment (jkozdon) [reply src/DGmethods/DGBalanceLawDiscretizations.jl]: I've added a `reverse_indefinite_stack_integral!` function

**Comment by jkozdon** [comment]
I've added a reverse function for the integral. Does this work for you @asraero and @smarras79?

**Pull request merged**